src/libraries/System.Private.CoreLib/src/System/Number.Dragon4.cs

```diff
@@ -100,203 +101,123 @@ public static unsafe void Dragon4Single(float value, int cutoffNumber, bool isSi
-                    BigInteger.Pow2((uint)(exponent + 1), out optionalMarginHigh);
-                }
-                else                // We have a fractional exponent
-                {
-                    // In order to track the mantissa data as an integer, we store it as is with a large scale
-
-                    // scaledValue      = 2 * 2 * mantissa
-                    BigInteger.SetUInt64(out scaledValue, 4 * mantissa);
-
-                    // scale            = 2 * 2 * 2^(-exponent)
-                    BigInteger.Pow2((uint)(-exponent + 2), out scale);
-
-                    // scaledMarginLow  = 2 * 2^(-1)
-                    BigInteger.SetUInt32(out scaledMarginLow, 1);
-
-                    // scaledMarginHigh = 2 * 2 * 2^(-1)
-                    BigInteger.SetUInt32(out optionalMarginHigh, 2);
-                }
-
-                // The high and low margins are different
-                pScaledMarginHigh = &optionalMarginHigh;
+                mantissaHighBitIdx = DiyFp.DoubleImplicitBitIndex;
+                hasUnequalMargins = (mantissa == (1UL << DiyFp.DoubleImplicitBitIndex));
             }
             else
             {
-                if (exponent > 0)   // We have no fractional component
-                {
-                    // 1) Expand the input value by multiplying out the mantissa and exponent.
-                    //    This represents the input value in its whole number representation.
-                    // 2) Apply an additional scale of 2 such that later comparisons against the margin values are simplified.
-                    // 3) Set the margin value to the lowest mantissa bit's scale.
-
-                    // scaledValue     = 2 * mantissa*2^exponent
-                    BigInteger.SetUInt64(out scaledValue, 2 * mantissa);
-                    scaledValue.ShiftLeft((uint)(exponent));
+                Debug.Assert(mantissa != 0);
+                mantissaHighBitIdx = (uint)BitOperations.Log2(mantissa);
+            }
 
-                    // scale           = 2 * 1
-                    BigInteger.SetUInt32(out scale, 2);
+            Dragon4State state = Dragon4GetScaleValueMargin(mantissa, exponent, mantissaHighBitIdx, hasUnequalMargins, cutoffNumber, isSignificantDigits);
 
-                    // scaledMarginLow = 2 * 2^(exponent-1)
-                    BigInteger.Pow2((uint)(exponent), out scaledMarginLow);
-                }
-                else                // We have a fractional exponent
-                {
-                    // In order to track the mantissa data as an integer, we store it as is with a large scale
+            Debug.Assert(state.scaledValue.GetLength() < 5);
+            // scaledValue is scaled up by 2 by the Dragon4 algorithm, undo that
+            BigInteger.SetUInt32(out BigInteger Two, 2);
+            BigInteger.DivRem(ref state.scaledValue, ref Two, out BigInteger descaledValue, out _); // TODO this could be optimized as a right shift
```

Comment:
I thought that `state.scaledValue` could be transformed into something usable by undoing the 2x scale Dragon4 applied, but I think there is other scaling I'm not accounting for.